### ISSUE TITLE
Flickering when zooming with terrain enabled

### ISSUE THREAD

**Issue by florianf** [maplibre/maplibre-gl-js#1077]
Hi all,

great work on the 3d feature, looks beautiful!

When testing your branch I noticed some flickering when zooming in (using the toursprung/terrain.json style) which is quite noticeable. It seems like semi-transparent polygons are rendered twice and for a few frames and the opacity is therefore doubled.
**Edit:** It only seems to happen on the first zooming of an area, when all tiles are cached it doesn't happen.

**Tile in the southeast with wrong opacity:**
![2200ms](https://user-images.githubusercontent.com/396450/154809201-6005db4c-338d-45be-93fb-6a6024f994fb.png)

**200ms later the tile is rendered correctly:**
![2400ms](https://user-images.githubusercontent.com/396450/154809249-911386a4-535d-4da6-bbae-5d99c99767e4.png)

This doesn't happen with terrain disabled, I think it's related to some networking activity (new DEM or image tiles available).

Any idea where that may come from? Happy to help if you can point me in the right direction.

Regards,
Florian

Original issue in prozessor13's repo: https://github.com/prozessor13/maplibre-gl-js/issues/60

**Comment by florianf** [comment]
I made some further tests. My initial suspicion was, it's because of the hillshade raster layer, but that's not the case. Confirmed this on another machine as well. Is this reproducible for anybody else? 

**Zooming in:**
![Screenshot_20220308_125019](https://user-images.githubusercontent.com/396450/157233431-39db9121-2a55-49b0-90dd-934bd75cd50f.png)

**First artifacts appear:**
![Screenshot_20220308_125041](https://user-images.githubusercontent.com/396450/157233484-2471f8b7-b10c-4fc0-a294-086e200571a8.png)

**Then some more:**
![Screenshot_20220308_125054](https://user-images.githubusercontent.com/396450/157233523-868632f1-7f0f-4096-8258-3df6cd0d3bae.png)

**When everything is loaded, everything is allright:**
![Screenshot_20220308_125118](https://user-images.githubusercontent.com/396450/157233583-ae929e36-5897-4dc5-8593-5b3a275d8a83.png)

Any pointers appreciated!

**Comment by prozessor13** [comment]
Hi, i think the issue is only with raster-tiles, and in the terrain-testpage the hillshading is done via raster-tiles.

In 2d mode the raster-tiles are rendered on every frame, and between zoomleves and loading the rastertiles will fade into each other, but 3d works completele different.
The vectortiles will render into a texture and this texture will draped over the terrain-mesh. During zooming maplibre resample existing rastertiles while loading the new tiles in background. But because of performance the rendert-to-texture is done only when tile-data changes, so in between this time you see the artefacts (e.g. when resampled and already loaded tiles are mixed).
Maybe this logic can be optimized, but currently i have no time to focus on that.

**Comment by florianf** [comment]
hi,
thanks for your reply. My first suspicion also were raster tiles, but take a look at the screenshots from yesterday, this also happens if the raster hillshade layer is switched off and no raster tiles are used.

>  so in between this time you see the artefacts (e.g. when resampled and already loaded tiles are mixed)
Could you point me to the location in the source?

**Comment by prozessor13** [comment]
OK, you are right, i will have a look onto it!

**Comment by florianf** [comment]
After some more digging, I can reproduce the overlapping rendering now in a simple example with only the natural-wood layer.
It seems, that in higher zoom levels there are overlapping tiles. Like here, it seems there is one parent 13 zoom tile, covered with two child tiles in 14, but not all four:

![Screenshot_20220318_214019](https://user-images.githubusercontent.com/396450/159081191-da4d6fc1-e850-4afc-a87b-bc58e27a7654.png)

**Comment by prozessor13** [comment]
Ok thanks, i think i know now where to start!

I think the problem is when filling [_coordsDescendingInv](https://github.com/maplibre/maplibre-gl-js/blob/terrain3d/src/render/render_to_texture.ts#L46). May after the fill process (below line 56) we have to check that there are no parent-tiles in the to-render-tiles array.

**Comment by florianf** [comment]
You're right, there are loads of parents in the to-render array. That's clear, because some low zoom tiles (f.ex. 5) are always visible. But the problem really only shows on zooming in, until all requested tiles are loaded. A simple hacky fix reduces the flickering to almost not noticeable.
I added this in the inner loop in [_init](https://github.com/maplibre/maplibre-gl-js/blob/terrain3d/src/render/render_to_texture.ts#L51=)
```
                let parentPresent = false
                for (const parentTileID of tileIDs) {
                    if (parentTileID != tileID && tileID.isChildOf(parentTileID) && (tileID.overscaledZ - parentTileID.overscaledZ) < 3) {
                        parentPresent = true
                    }
                }
                
                if (parentPresent) {
                    continue;
                }
```
The tile isn't rendered, if there is a parent or grandparent tile visible. I don't get why the double rendering isn't notice all the time, because tiles are rendered on top of another all the time?

**Comment by prozessor13** [comment]
Thx for your effort! I shortly digged into this but currently i do not have a correct answer for the issue. As soon as i have time, i will look again onto this issue.

**Comment by florianf** [comment]
@prozessor13 I finally found the issue, took me a couple of hours. Based on your #1651 PR:

The problem is in the renderLayer methode in render_to_texture.ts. When a zoom event occurs and not all tiles are completely loaded, the coords array in the layer loop contains overlapping tiles. Link to source: https://github.com/prozessor13/maplibre-gl-js/blob/c82c13eac651bf96f23e44985bbf62326fc93ad2/src/render/render_to_texture.ts#L247

![grafik](https://user-images.githubusercontent.com/396450/191023935-5892bea0-81c2-4953-94b9-08b9a062ac61.png)

My suggested fix would be, remove the overlapping tiles for fill layers. Lower zoom levels should be always rendered first, because of ordering in coords array.

```typescript
                for (let l = 0; l < layers.length; l++) {
                    const layer = painter.style._layers[layers[l]];
                    let coords = layer.source ? this._coordsDescendingInv[layer.source][tile.tileID.key] : [tile.tileID];
                    if (coords && coords.length > 1 && layer.type == "fill") {
                        let parentPresent = false
                        let nonOverlappingCoords = []
                        for (const tileID of coords) {
                            for (const parentTileID of coords) {
                                if (parentTileID != tileID && tileID.isChildOf(parentTileID)) {
                                    parentPresent = true
                                    break;
                                }
                            }
                            if (!parentPresent) {
                                nonOverlappingCoords[nonOverlappingCoords.length] = tileID
                            }
                        }
                        coords = nonOverlappingCoords
                    }
                    painter._renderTileClippingMasks(layer, coords);
                    painter.renderLayer(painter, painter.style.sourceCaches[layer.source], layer, coords);
                    if (layer.source) tile.rttCoords[layer.source] = this._coordsDescendingInvStr[layer.source][tile.tileID.key];
                }
```
What do you think?

**Comment by florianf** [comment]
I tested your rtt_stencil branch an it works great! It also fixes another related issue where tile borders (overdraw) where overlapping when zooming out. Really fantastic!

**Comment by HarelM** [comment]
@florianf I've assigned a bounty for adding tests to the #1672 in #1540.
Feel free to add some test so the PR can be merged.

**Comment by HarelM** [comment]
No need to assign a bounty here since it was assigned in #1540.
Link to parent Bounty: https://github.com/maplibre/maplibre/issues/189

**Comment by clement-igonet** [comment]
The fix discussed here, the stencil logic for render-to-texture tiles, was merged in April 2023 (#1672).
@florianf confirmed above that the branch fixed the flickering, and the related border overdraw when zooming out.
Is there anything left to keep this issue open for, or can it be closed?

**Comment by HarelM** [comment]
Closing, if there's a need to open we can do that too.
